# NLP 08: Parse with Fuzzy Matches

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import display
from time import gmtime, strftime
import sys
import os
import io

import string
import re
# import itertools
# import nltk
# nltk.download('stopwords')

from fuzzywuzzy import fuzz
from fuzzywuzzy import process
from rapidfuzz import fuzz as rfuzz
import jaro

In [2]:
def frequency_ct(ngram_list):
    freq_dict = {}
    for ngram in ngram_list:
        if ngram not in freq_dict:
            freq_dict[ngram] = 0
        freq_dict[ngram] +=1
    return freq_dict

def calc_fuzz_df(df, column):
    row_list = []
    
    for o_i, o_v in enumerate(df[column].sort_index()):
        for m_i, m_v in enumerate(df[column].sort_index()):
            if o_i != m_i:
                dict1 = {
                    'original_index': o_i,
                    'original_value': o_v,
                    'match_index': m_i,
                    'match_value': m_v,
                    'ratio_score': rfuzz.ratio(o_v, m_v),
                    'partial_ratio_score': rfuzz.partial_ratio(o_v, m_v),
                    'token_sort_score': rfuzz.token_sort_ratio(o_v, m_v),
                    'token_set_score': rfuzz.token_set_ratio(o_v, m_v),
                    'jaro_winkler_score': jaro.jaro_winkler_metric(o_v, m_v)
                }
                if (dict1['ratio_score']>60) | (dict1['partial_ratio_score']>60) | (dict1['token_sort_score']>60) | (dict1['token_set_score']>60) | (dict1['jaro_winkler_score']>0.6):
                    row_list.append(dict1)
    score_df = pd.DataFrame(row_list)
        
    return score_df

In [3]:
df = pd.read_csv('data/parsed_bahamas_addresses.csv')

df['address_wordlist'] = df['working_address'].fillna('').str.split()

freq_df = pd.DataFrame.from_dict(
    frequency_ct(df['address_wordlist'].sum()
                ), orient='index').reset_index().rename(
    columns={'index':'word', 0:'count'}).sort_values('count', ascending=False)

In [4]:
freq_df.head(10)

word  count
9    bahamas   2324
8     nassau   2043
6        box   1484
5         po   1430
4     street   1128
2        and    627
3    shirley    489
10     suite    447
34       bay    431
14  building    329

In [5]:
fuzzy_words_df = calc_fuzz_df(freq_df, 'word')
fuzzy_words_df['jaro_winkler_score'] = fuzzy_words_df['jaro_winkler_score']*100

#### Cities

- Nassau
- Freeport
- Marsh Harbour
- Governor's Harbour
- Windermere island
- Harbour Island
- Elbow Cay
- Treasure Cay
- Gregory Town
- Spanish Wells

#### Islands

- New Providence
- Paradise Island
- Grand Bahama
- Abaco
- Eleuthera
- South Andros

In [6]:
df['address_wordlist'].str[-1].value_counts()[:20]

bahamas       2043
bahama          66
nassau          54
freeport         5
street           5
bahmas           5
bahams           5
i                4
providence       3
centre           3
343              3
lane             3
kelty            2
bahamas1         2
kln              2
2423527291       2
bhs              2
abaco            2
isle             1
esquare          1
Name: address_wordlist, dtype: int64

## Bahamas

Remove "bahamas" and the misspellings from the last word in the word list. I'm removing the last word only so that I don't accidentally take "bahamas" out of phrases like "Bahamas Financial Centre." I want to try to separate out such building names later.

This lists the "bahamas" matches from the fuzzy matching

In [7]:
bah_fuzz_list = fuzzy_words_df[(fuzzy_words_df['original_value']=='bahamas') & (fuzzy_words_df['ratio_score']>75)].sort_values(['ratio_score', 'original_value'], ascending=False)['match_value'].to_list()
bah_fuzz_list

['bahamasc',
 'bahamas1',
 'bahamaas',
 'bahamas6',
 'bahamasa',
 'abahamas',
 'bahama',
 'bahams',
 'bhamas',
 'bahmas',
 'ahamas',
 'bahamaspo',
 'bahaams',
 'bahanas',
 'brahmas',
 'baham',
 'hamas']

These are the matches I can get with a regular expression on the last word of the address word list.

In [8]:
bah_reg_list = df.loc[df['working_address'].str.contains(r'ba?h\w+s\w?\d?$', regex=True), 'address_wordlist'].str[-1].unique().tolist()
bah_reg_list

['bahamas',
 'bahams',
 'bhamas',
 'bahmas',
 'bahamasc',
 'bahamas1',
 'bahaams',
 'bahamaas',
 'bahanas',
 'bahamas6',
 'bahamasa',
 'abahamas']

One thing to keep in mind is that the fuzzy matches are from all unique words, which means not all of them are necessarily going to be present in the last word.

In [9]:
set(bah_fuzz_list) - set(bah_reg_list)

{'ahamas', 'baham', 'bahama', 'bahamaspo', 'brahmas', 'hamas'}

In [10]:
set(bah_reg_list) - set(bah_fuzz_list) 

{'bahamas'}

In [11]:
pd.set_option('display.max_colwidth', 1000)

In [12]:
df['address_wordlist'].apply(lambda x: x[:-1] if x[-1] in bah_reg_list else x)

0                                                                                    [annex, frederick, and, shirley, street, po, box, n4805, nassau]
1                                                                                         [suite, e2, union, court, building, po, box, n8188, nassau]
2                                                                                           [lyford, cay, house, lyford, cay, po, box, n7785, nassau]
3                                               [po, box, n3708, bahamas, financial, centre, po, box, n3708, shirley, and, charlotte, street, nassau]
4                                                                             [lyford, cay, house, third, floor, lyford, cay, po, box, n3024, nassau]
                                                                            ...                                                                      
2253                                                                            [j, p, morgan, trust

In [13]:
df['address_country'] = df['address_wordlist'].apply(lambda x:'bahamas' if x[-1] in bah_reg_list else np.nan)
df['address_wordlist'] = df['address_wordlist'].apply(lambda x: x[:-1] if x[-1] in bah_reg_list else x)

In [14]:
df.head()

node_id  \
0  24000001   
1  24000002   
2  24000003   
3  24000004   
4  24000005   

                                                                                              address  \
0                                     ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                    SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                      LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3  P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                           LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   

  name countries country_codes       sourceID  \
0  NaN   Bahamas           BHS  Bahamas Leaks   
1  NaN   Bahamas           BHS  Bahamas Leaks   
2  NaN   Bahamas           BHS  Bahamas Leaks   
3  NaN   Bahamas           BHS  Bahamas Leaks   
4  NaN   Bahamas           BHS  Bahamas Leaks   

                                             valid_until note  \
0  The Bahamas Leaks data is current through early 2016.  NaN   
1  The Bahamas Leaks data is current through early 2016.  NaN   
2  The Bahamas Leaks data is current through early 2016.  NaN   
3  The Bahamas Leaks data is current through early 2016.  NaN   
4  The Bahamas Leaks data is current through early 2016.  NaN   

                                                                                  working_address  \
0                                  annex frederick and shirley street po box n4805 nassau bahamas   
1                                       suite e2 union court building po box n8188 nassau bahamas   
2                                         lyford cay house lyford cay po box n7785 nassau bahamas   
3  po box n3708 bahamas financial centre po box n3708 shirley and charlotte street nassau bahamas   
4                             lyford cay house third floor lyford cay po box n3024 nassau bahamas   

                                                                                        address_wordlist  \
0                                       [annex, frederick, and, shirley, street, po, box, n4805, nassau]   
1                                            [suite, e2, union, court, building, po, box, n8188, nassau]   
2                                              [lyford, cay, house, lyford, cay, po, box, n7785, nassau]   
3  [po, box, n3708, bahamas, financial, centre, po, box, n3708, shirley, and, charlotte, street, nassau]   
4                                [lyford, cay, house, third, floor, lyford, cay, po, box, n3024, nassau]   

  address_country  
0         bahamas  
1         bahamas  
2         bahamas  
3         bahamas  
4         bahamas

## The

"the" shows up a surprising amount. It probably isn't relevant, but I prefer to double check.

In [15]:
df['address_wordlist'].str[-1].value_counts()

nassau        1815
providence      78
bahama          76
the             69
freeport        16
              ... 
n08188           1
square           1
islands          1
ba               1
cb12399          1
Name: address_wordlist, Length: 110, dtype: int64

I can see that when "the" is the second to last word it follows "bahamas" as in "the bahamas." As such, I can safely remove it.

In [16]:
pd.set_option('display.max_rows', 70)

In [17]:
df['working_address'].str.split().apply(lambda x: x[-2:]).value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 5231, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[nassau, bahamas]        1743
[providence, bahamas]      74
[the, bahamas]             68
[grand, bahama]            54
[abaco, bahamas]           13
                         ... 
[n, 8188]                   1
[nassau, bahaams]           1
[343nassau, bahamas]        1
[generals, dept]            1
[cb12399, bahamas]          1
Name: working_address, Length: 176, dtype: int64

After removing the "bahamas" like values, there is now at least one empty list. From here on out I'll need to account for empty lists.

In [18]:
df['address_wordlist'] = df['address_wordlist'].apply(lambda x: x if not x else x[:-1] if x[-1] in ['the'] else x)

In [19]:
df['address_wordlist'].str[-1].value_counts()

nassau        1854
providence      89
bahama          77
freeport        16
abaco           15
              ... 
n08188           1
square           1
islands          1
ba               1
cb12399          1
Name: address_wordlist, Length: 111, dtype: int64

## Nassau

In [20]:
fuzzy_words_df[(fuzzy_words_df['original_value']=='nassau') & (fuzzy_words_df['ratio_score']>70)].sort_values(['ratio_score', 'match_value'], ascending=False)

original_index original_value  match_index match_value  ratio_score  \
1236               8         nassau         1593     naussau    92.307692   
1188               8         nassau          698     nasssau    92.307692   
1220               8         nassau         1224     nassaus    92.307692   
1198               8         nassau          925     nassaub    92.307692   
1214               8         nassau         1160     nassaau    92.307692   
1216               8         nassau         1194       nassu    90.909091   
1196               8         nassau          872       nasau    90.909091   
1212               8         nassau         1146      nassua    83.333333   
1215               8         nassau         1185      nassao    83.333333   
1217               8         nassau         1201      nassan    83.333333   
1209               8         nassau         1083      massau    83.333333   
1203               8         nassau          993   343nassau    80.000000   
1221               8         nassau         1243       massa    72.727273   

      partial_ratio_score  token_sort_score  token_set_score  \
1236            83.333333         92.307692        92.307692   
1188            83.333333         92.307692        92.307692   
1220           100.000000         92.307692        92.307692   
1198           100.000000         92.307692        92.307692   
1214            90.909091         92.307692        92.307692   
1216            88.888889         90.909091        90.909091   
1196            80.000000         90.909091        90.909091   
1212            90.909091         83.333333        83.333333   
1215            90.909091         83.333333        83.333333   
1217            90.909091         83.333333        83.333333   
1209            90.909091         83.333333        83.333333   
1203           100.000000         80.000000        80.000000   
1221            80.000000         72.727273        72.727273   

      jaro_winkler_score  
1236           96.190476  
1188           97.142857  
1220           97.142857  
1198           97.142857  
1214           97.142857  
1216           96.666667  
1196           96.111111  
1212           96.666667  
1215           93.333333  
1217           93.333333  
1209           88.888889  
1203           88.888889  
1221           82.222222

In [21]:
nas_ratio_list = fuzzy_words_df.loc[(fuzzy_words_df['original_value']=='nassau') & (fuzzy_words_df['ratio_score']>70), 'match_value'].to_list()
nas_jaro_list = fuzzy_words_df.loc[(fuzzy_words_df['original_value']=='nassau') & (fuzzy_words_df['jaro_winkler_score']>85), 'match_value'].to_list()
pd.DataFrame([nas_ratio_list, nas_jaro_list]).T

0          1
0     nasssau    nasssau
1       nasau      nasau
2     nassaub    nassaub
3   343nassau  343nassau
4      massau     massau
5      nassua     nassua
6     nassaau    nassaau
7      nassao     nassao
8       nassu      nassu
9      nassan     nassan
10    nassaus    nassaus
11      massa    naussau
12    naussau       None

### Setting the threshold

The options above 79 look good for the `ratio` look good. In examining the values in the data, I see that "massa" is actually an address in Italy and I'll need to separate "343nassau" into two parts to retain the rest of the address.

In [22]:
for word in nas_ratio_list:
    print(word)
    display(df[df['address_wordlist'].apply(lambda x: word in x)])
    print('\n')

nasssau


node_id  \
420  24000421   

                                                                      address  \
420  3RD FLOOR, GEORGE HOUSE, GEORGE STREET, P.O. BOX N-8159 NASSSAU, BAHAMAS   

    name countries country_codes       sourceID  \
420  NaN   Bahamas           BHS  Bahamas Leaks   

                                               valid_until note  \
420  The Bahamas Leaks data is current through early 2016.  NaN   

                                                         working_address  \
420  third floor george house george street po box n8159 nasssau bahamas   

                                                           address_wordlist  \
420  [third, floor, george, house, george, street, po, box, n8159, nasssau]   

    address_country  
420         bahamas



nasau


node_id  \
536   14000678   
612   14030188   
682   14035228   
724   14038327   
965   14078960   
1440    239867   

                                                                                           address  \
536                                                            101 East Hill Street, Nasau Bahamas   
612                Bahamas Financial Centre; Shirley & Charlotte Street; Fourth Flor Nasau Bahamas   
682                                                                       CB 11-343 Nasau, Bahamas   
724   Elizabeth Avenue and Shirley Street; Union Court Building; Suite E-2; N-8188; Nasau; Bahamas   
965                        Suite 102; Saffrey Square; Bay Street and Bank Lane; Nasau; The Bahamas   
1440      UBS Trustees (Bahamas) Ltd, UBS House, East Bay Street, P. O. Box N-7757, Nasau, Bahamas   

     name countries country_codes        sourceID  \
536   NaN   Bahamas           BHS   Panama Papers   
612   NaN   Bahamas           BHS   Panama Papers   
682   NaN   Bahamas           BHS   Panama Papers   
724   NaN   Bahamas           BHS   Panama Papers   
965   NaN   Bahamas           BHS   Panama Papers   
1440  NaN   Bahamas           BHS  Offshore Leaks   

                                          valid_until note  \
536   The Panama Papers  data is current through 2015  NaN   
612   The Panama Papers  data is current through 2015  NaN   
682   The Panama Papers  data is current through 2015  NaN   
724   The Panama Papers  data is current through 2015  NaN   
965   The Panama Papers  data is current through 2015  NaN   
1440  The Offshore Leaks data is current through 2010  NaN   

                                                                              working_address  \
536                                                        101 east hill street nasau bahamas   
612           bahamas financial centre shirley and charlotte street fourth flor nasau bahamas   
682                                                                   cb 11 343 nasau bahamas   
724   elizabeth avenue and shirley street union court building suite e 2 n 8188 nasau bahamas   
965                       suite 102 saffrey square bay street and bank lane nasau the bahamas   
1440           ubs trustees bahamas ltd ubs house east bay street p o box n7757 nasau bahamas   

                                                                                    address_wordlist  \
536                                                                 [101, east, hill, street, nasau]   
612               [bahamas, financial, centre, shirley, and, charlotte, street, fourth, flor, nasau]   
682                                                                             [cb, 11, 343, nasau]   
724   [elizabeth, avenue, and, shirley, street, union, court, building, suite, e, 2, n, 8188, nasau]   
965                               [suite, 102, saffrey, square, bay, street, and, bank, lane, nasau]   
1440           [ubs, trustees, bahamas, ltd, ubs, house, east, bay, street, p, o, box, n7757, nasau]   

     address_country  
536          bahamas  
612          bahamas  
682          bahamas  
724          bahamas  
965          bahamas  
1440         bahamas



nassaub


node_id  \
583  14026897   

                                                                                    address  \
583  ANSBACHER (BAHAMAS) LIMITED P.O. BOX N 7768 ANSBACHER HOUSE BANK LANE NASSAUB, BAHAMAS   

    name countries country_codes       sourceID  \
583  NaN   Bahamas           BHS  Panama Papers   

                                         valid_until note  \
583  The Panama Papers  data is current through 2015  NaN   

                                                                       working_address  \
583  ansbacher bahamas limited po box n 7768 ansbacher house bank lane nassaub bahamas   

                                                                           address_wordlist  \
583  [ansbacher, bahamas, limited, po, box, n, 7768, ansbacher, house, bank, lane, nassaub]   

    address_country  
583         bahamas



343nassau


node_id                   address name countries country_codes  \
681  14035227  CB 11.343/Nassau Bahamas  NaN   Bahamas           BHS   

          sourceID                                      valid_until note  \
681  Panama Papers  The Panama Papers  data is current through 2015  NaN   

             working_address     address_wordlist address_country  
681  cb 11 343nassau bahamas  [cb, 11, 343nassau]         bahamas



massau


node_id          address name countries country_codes       sourceID  \
810  14049672  MASSAU, BAHAMAS  NaN   Bahamas           BHS  Panama Papers   

                                         valid_until note working_address  \
810  The Panama Papers  data is current through 2015  NaN  massau bahamas   

    address_wordlist address_country  
810         [massau]         bahamas



nassua


node_id  \
933  14077075   
969  14078964   

                                                                      address  \
933    SAFFREY SQUARE; SUITE 205; BANK LANE; P.O. BOX N-8188; NASSUA; BAHAMAS   
969  Suite 102; Saffrey Square; Bay Street and Bank Lane; Nassua; The Bahamas   

    name countries country_codes       sourceID  \
933  NaN   Bahamas           BHS  Panama Papers   
969  NaN   Bahamas           BHS  Panama Papers   

                                         valid_until note  \
933  The Panama Papers  data is current through 2015  NaN   
969  The Panama Papers  data is current through 2015  NaN   

                                                          working_address  \
933        saffrey square suite 205 bank lane po box n8188 nassua bahamas   
969  suite 102 saffrey square bay street and bank lane nassua the bahamas   

                                                        address_wordlist  \
933    [saffrey, square, suite, 205, bank, lane, po, box, n8188, nassua]   
969  [suite, 102, saffrey, square, bay, street, and, bank, lane, nassua]   

    address_country  
933         bahamas  
969         bahamas



nassaau


node_id  \
966  14078961   

                                                                      address  \
966  Suite 102; Saffrey Square; Bay Street and Bank Lane Nassaau; The Bahamas   

    name countries country_codes       sourceID  \
966  NaN   Bahamas           BHS  Panama Papers   

                                         valid_until note  \
966  The Panama Papers  data is current through 2015  NaN   

                                                           working_address  \
966  suite 102 saffrey square bay street and bank lane nassaau the bahamas   

                                                         address_wordlist  \
966  [suite, 102, saffrey, square, bay, street, and, bank, lane, nassaau]   

    address_country  
966         bahamas



nassao


node_id  \
999  14079956   

                                                                                                    address  \
999  Suite E-2; Union Court Buiding; Elizabeth Avenue and Shirley Streer; P.O. Box N-8188; Nassao; Bahamas.   

    name countries country_codes       sourceID  \
999  NaN   Bahamas           BHS  Panama Papers   

                                         valid_until note  \
999  The Panama Papers  data is current through 2015  NaN   

                                                                                   working_address  \
999  suite e2 union court buiding elizabeth avenue and shirley streer po box n8188 nassao bahamas    

                                                                                        address_wordlist  \
999  [suite, e2, union, court, buiding, elizabeth, avenue, and, shirley, streer, po, box, n8188, nassao]   

    address_country  
999         bahamas



nassu


node_id  \
1021  14079979   
1117  14080667   

                                                                                           address  \
1021            Suite E-2; Union Court Building; Elizabeth Avenue & Shirley Street; Nassu; Bahamas   
1117  The Bahamas Financial Centre; Shirley and Charlotte Streets; P.O. Box N-3023; Nassu; Bahamas   

     name countries country_codes       sourceID  \
1021  NaN   Bahamas           BHS  Panama Papers   
1117  NaN   Bahamas           BHS  Panama Papers   

                                          valid_until note  \
1021  The Panama Papers  data is current through 2015  NaN   
1117  The Panama Papers  data is current through 2015  NaN   

                                                                           working_address  \
1021      suite e 2 union court building elizabeth avenue and shirley street nassu bahamas   
1117  the bahamas financial centre shirley and charlotte street po box n3023 nassu bahamas   

                                                                               address_wordlist  \
1021      [suite, e, 2, union, court, building, elizabeth, avenue, and, shirley, street, nassu]   
1117  [the, bahamas, financial, centre, shirley, and, charlotte, street, po, box, n3023, nassu]   

     address_country  
1021         bahamas  
1117         bahamas



nassan


node_id  \
1043  14080003   
1050  14080011   

                                                                                                 address  \
1043  Suite E-2; Union Court Building; Elizabeth Avenue and Shirley Street; PO Box 8188; Nassan; Bahamas   
1050            Suite E-2; Union Court Building; Elizabeth Avenue and Shirly Street; Nassan; The Bahamas   

     name countries country_codes       sourceID  \
1043  NaN   Bahamas           BHS  Panama Papers   
1050  NaN   Bahamas           BHS  Panama Papers   

                                          valid_until note  \
1043  The Panama Papers  data is current through 2015  NaN   
1050  The Panama Papers  data is current through 2015  NaN   

                                                                                    working_address  \
1043  suite e 2 union court building elizabeth avenue and shirley street po box 8188 nassan bahamas   
1050           suite e 2 union court building elizabeth avenue and shirly street nassan the bahamas   

                                                                                           address_wordlist  \
1043  [suite, e, 2, union, court, building, elizabeth, avenue, and, shirley, street, po, box, 8188, nassan]   
1050                  [suite, e, 2, union, court, building, elizabeth, avenue, and, shirly, street, nassan]   

     address_country  
1043         bahamas  
1050         bahamas



nassaus


node_id  \
1106  14080656   

                                                                                         address  \
1106  The Bahamas Financial Centre; Shirley & Charlotte Streets; PO Box N-3023; Nassaus; Bahamas   

     name countries country_codes       sourceID  \
1106  NaN   Bahamas           BHS  Panama Papers   

                                          valid_until note  \
1106  The Panama Papers  data is current through 2015  NaN   

                                                                             working_address  \
1106  the bahamas financial centre shirley and charlotte street po box n3023 nassaus bahamas   

                                                                                 address_wordlist  \
1106  [the, bahamas, financial, centre, shirley, and, charlotte, street, po, box, n3023, nassaus]   

     address_country  
1106         bahamas



massa


node_id                        address name countries country_codes  \
1150  14083823  VIA BIGINI; 43; I-54100 MASSA  NaN   Bahamas           BHS   

           sourceID                                      valid_until note  \
1150  Panama Papers  The Panama Papers  data is current through 2015  NaN   

                  working_address                    address_wordlist  \
1150  via bigini 43 i 54100 massa  [via, bigini, 43, i, 54100, massa]   

     address_country  
1150             NaN



naussau


node_id                                                address name  \
1428   252371  43 Elizabeth Avenue, P.O.Box CB-13022 Naussau Bahamas  NaN   

     countries country_codes        sourceID  \
1428   Bahamas           BHS  Offshore Leaks   

                                          valid_until note  \
1428  The Offshore Leaks data is current through 2010  NaN   

                                         working_address  \
1428  43 elizabeth avenue po box cb13022 naussau bahamas   

                                        address_wordlist address_country  
1428  [43, elizabeth, avenue, po, box, cb13022, naussau]         bahamas

I went back and updated the original standardization. I left the example where I found it.

In [23]:
df.loc[df['working_address'].str.contains("343nass"), 'working_address'] = 'cb 11 343 nassau bahamas'

In [24]:
nas_fuzz_list = fuzzy_words_df.loc[(fuzzy_words_df['original_value']=='nassau') & (fuzzy_words_df['ratio_score']>80), 'match_value'].to_list()
nas_fuzz_list = nas_fuzz_list + ['nassau']
nas_fuzz_list

['nasssau',
 'nasau',
 'nassaub',
 'massau',
 'nassua',
 'nassaau',
 'nassao',
 'nassu',
 'nassan',
 'nassaus',
 'naussau',
 'nassau']

In [25]:
df['address_city'] = df['address_wordlist'].apply(lambda x: np.nan if not x else 'nassau' if x[-1] in nas_fuzz_list else np.nan)
df['address_wordlist'] = df['address_wordlist'].apply(lambda x: x if not x else x[:-1] if x[-1] in nas_fuzz_list else x)

In [26]:
df[df['address_city']=='nassau']

node_id  \
0      24000001   
1      24000002   
2      24000003   
3      24000004   
4      24000005   
...         ...   
2226  240491372   
2227  240491474   
2229  240491518   
2248  240492221   
2255  240491733   

                                                                                                                           address  \
0                                                                  ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                                                 SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                                                   LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3                               P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                                                        LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   
...                                                                                                                            ...   
2226                                                                    MONTAGUE STERLING CENTRE, EAST BAY STREET, NASSAU, BAHAMAS   
2227                                     SUITE 200B, 2ND FLOOR, CENTRE OF COMMERCE, ONE BAY STREET, PO BOX N-3944, NASSAU, BAHAMAS   
2229                   RBC TRUST COMPANY (BAHAMAS) LIMITED, BAYSIDE EXECUTIVE PARK BUILDING 3, P.O. BOX NO. 30-24, NASSAU, BAHAMAS   
2248  JPMORGAN TRUST COMPANY (BAHAMAS) LIMITED, 2ND FLOOR, BAHAMAS FINANCIAL CENTRE, SHIRLEY AND CHARLOTTE STREET, NASSAU, BAHAMAS   
2255                                                                      DELTEC HOUSE, LYFORD CAY, PO BOX N-3229, NASSAU, BAHAMAS   

     name countries country_codes                        sourceID  \
0     NaN   Bahamas           BHS                   Bahamas Leaks   
1     NaN   Bahamas           BHS                   Bahamas Leaks   
2     NaN   Bahamas           BHS                   Bahamas Leaks   
3     NaN   Bahamas           BHS                   Bahamas Leaks   
4     NaN   Bahamas           BHS                   Bahamas Leaks   
...   ...       ...           ...                             ...   
2226  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2227  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2229  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2248  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2255  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   

                                                valid_until note  \
0     The Bahamas Leaks data is current through early 2016.  NaN   
1     The Bahamas Leaks data is current through early 2016.  NaN   
2     The Bahamas Leaks data is current through early 2016.  NaN   
3     The Bahamas Leaks data is current through early 2016.  NaN   
4     The Bahamas Leaks data is current through early 2016.  NaN   
...                                                     ...  ...   
2226                  Provider data is current through 2016  NaN   
2227                  Provider data is current through 2016  NaN   
2229                  Provider data is current through 2016  NaN   
2248                  Provider data is current through 2017  NaN   
2255                  Provider data is current through 2016  NaN   

                                                                                                               working_address  \
0                                                               annex frederick and shirley street po box n4805 nassau bahamas   
1                                                                    suite e2 union court building po box n8188 nassau bahamas   
2                                                                      lyford cay house lyford cay po box n7785 nassau bahamas   
3   

## Next step

The value counts for the last word are now heavily weighted toward PO Box numbers. "n8188" and "8188" refer to the same PO Box. I believe the "cb" and "ss" entries are also PO Boxes.

I'm going to join the working address back into a single string and pull out the PO Boxes.

In [27]:
df['address_wordlist'].str[-1].value_counts()

street        434
n8188         109
providence     93
bahama         77
8188           54
             ... 
cb11103         1
ss19772         1
1462            1
cb13908         1
zh              1
Name: address_wordlist, Length: 665, dtype: int64

In [28]:
df['address_wordlist'] = df['address_wordlist'].apply(lambda x: '' if not x else ' '.join(x))
df

node_id  \
0      24000001   
1      24000002   
2      24000003   
3      24000004   
4      24000005   
...         ...   
2253  240492525   
2254  240492536   
2255  240491733   
2256  240491778   
2257  240491892   

                                                                                                                                          address  \
0                                                                                 ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                                                                SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                                                                  LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3                                              P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                                                                       LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   
...                                                                                                                                           ...   
2253                                                                      J.P.MORGAN TRUST COMPANY (BAHAMAS) LIMITED, NASSAU, N-4899, ZH, BAHAMAS   
2254                                                                  MONTAGNE STERLINE CENTRE. EAST BAV STREET, NASSAU, COUNTRY BAHAMAS, BAHAMAS   
2255                                                                                     DELTEC HOUSE, LYFORD CAY, PO BOX N-3229, NASSAU, BAHAMAS   
2256                                                  PROVIDENCE HOUSE, HAST WING, EAST HILL STREET, P.O. BOX CB-12399, NASSAU, CB-12399, BAHAMAS   
2257  C/OJ.P. MORGAN TRUST COMPANY (BAHAMAS) LIMITED, BAHAMAS FINANCIAL CENTER 2NTL FLOOR, SHIRLEY AND CHARLOTTE STREETS, NASSAU, N-4899, BAHAMAS   

     name countries country_codes                        sourceID  \
0     NaN   Bahamas           BHS                   Bahamas Leaks   
1     NaN   Bahamas           BHS                   Bahamas Leaks   
2     NaN   Bahamas           BHS                   Bahamas Leaks   
3     NaN   Bahamas           BHS                   Bahamas Leaks   
4     NaN   Bahamas           BHS                   Bahamas Leaks   
...   ...       ...           ...                             ...   
2253  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2254  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2255  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2256  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2257  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   

                                                valid_until note  \
0     The Bahamas Leaks data is current through early 2016.  NaN   
1     The Bahamas Leaks data is current through early 2016.  NaN   
2     The Bahamas Leaks data is current through early 2016.  NaN   
3     The Bahamas Leaks data is current through early 2016.  NaN   
4     The Bahamas Leaks data is current through early 2016.  NaN   
...                                                     ...  ...   
2253                  Provider data is current through 2016  NaN   
2254                  Provider data is current through 2016  NaN   
2255                  Provider data is current through 2016  NaN   
2256                  Provider data is current through 2016  NaN   
2257                  Provider data is current through 2017  NaN   

                                                                                                                        working_address  \
0                                                                        annex frederick and shirley street po box n4805 nassau bahamas   
1                                                                 

In [30]:
df[df['address_wordlist'].str.contains('box', regex=True)]

node_id  \
0      24000001   
1      24000002   
2      24000003   
3      24000004   
4      24000005   
...         ...   
2225  240491356   
2227  240491474   
2229  240491518   
2255  240491733   
2256  240491778   

                                                                                                            address  \
0                                                   ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                                  SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                                    LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3                P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                                         LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   
...                                                                                                             ...   
2225  P.O. BOX N- 3944, SUITE 200B, 2ND FLOOR, CENTRE OF COMMERCE, ONE BAY STREET, NASSAU, BAHAMAS, NASSAU, BAHAMAS   
2227                      SUITE 200B, 2ND FLOOR, CENTRE OF COMMERCE, ONE BAY STREET, PO BOX N-3944, NASSAU, BAHAMAS   
2229    RBC TRUST COMPANY (BAHAMAS) LIMITED, BAYSIDE EXECUTIVE PARK BUILDING 3, P.O. BOX NO. 30-24, NASSAU, BAHAMAS   
2255                                                       DELTEC HOUSE, LYFORD CAY, PO BOX N-3229, NASSAU, BAHAMAS   
2256                    PROVIDENCE HOUSE, HAST WING, EAST HILL STREET, P.O. BOX CB-12399, NASSAU, CB-12399, BAHAMAS   

     name countries country_codes                        sourceID  \
0     NaN   Bahamas           BHS                   Bahamas Leaks   
1     NaN   Bahamas           BHS                   Bahamas Leaks   
2     NaN   Bahamas           BHS                   Bahamas Leaks   
3     NaN   Bahamas           BHS                   Bahamas Leaks   
4     NaN   Bahamas           BHS                   Bahamas Leaks   
...   ...       ...           ...                             ...   
2225  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2227  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2229  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2255  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2256  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   

                                                valid_until note  \
0     The Bahamas Leaks data is current through early 2016.  NaN   
1     The Bahamas Leaks data is current through early 2016.  NaN   
2     The Bahamas Leaks data is current through early 2016.  NaN   
3     The Bahamas Leaks data is current through early 2016.  NaN   
4     The Bahamas Leaks data is current through early 2016.  NaN   
...                                                     ...  ...   
2225                  Provider data is current through 2017  NaN   
2227                  Provider data is current through 2016  NaN   
2229                  Provider data is current through 2016  NaN   
2255                  Provider data is current through 2016  NaN   
2256                  Provider data is current through 2016  NaN   

                                                                                            working_address  \
0                                            annex frederick and shirley street po box n4805 nassau bahamas   
1                                                 suite e2 union court building po box n8188 nassau bahamas   
2                                                   lyford cay house lyford cay po box n7785 nassau bahamas   
3            po box n3708 bahamas financial centre po box n3708 shirley and charlotte street nassau bahamas   
4                                       lyford cay house third floor lyford cay po box n3024 nassau bahamas   
...                                   

In [47]:
df.loc[df['address_wordlist'].str.contains('p?o?\s?box\s?\w+\s?\d+', regex=True), 'address_wordlist']

0                                              annex frederick and shirley street po box n4805
1                                                   suite e2 union court building po box n8188
2                                                     lyford cay house lyford cay po box n7785
3              po box n3708 bahamas financial centre po box n3708 shirley and charlotte street
4                                         lyford cay house third floor lyford cay po box n3024
                                                 ...                                          
2225    po box n 3944 suite 200b second floor centre of commerce one bay street nassau bahamas
2227                    suite 200b second floor centre of commerce one bay street po box n3944
2229       rbc trust company bahamas limited bayside executive park building 3 po box no 30 24
2255                                                      deltec house lyford cay po box n3229
2256                 providence house hast wing ea

In [132]:
pattern2 = '(p?o?\s?b?o?x?\s?box\s?\w?\s?\w+\s?\d+)'
df['po_box'] = df['address_wordlist'].str.extract(pattern2)
df['address_wordlist'] = df['address_wordlist'].str.replace(pattern2, '', regex=True)

df['address_wordlist'] = df['address_wordlist'].str.replace('(po box)', '', regex=True)

pattern = r'(\bpo\s.+\s?\d+)'
df.loc[df['address_wordlist'].str.contains(pattern), 'po_box'
      ] = df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'].str.extract(pattern, expand=False)
df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'
      ] = df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'].str.replace(pattern, '', regex=True)

pattern1 = r'(\bpo\s?b\s?\w\s?\d+\b)'
df.loc[df['address_wordlist'].str.contains(pattern1), 'po_box'
      ] = df.loc[df['address_wordlist'].str.contains(pattern1), 'address_wordlist'].str.extract(pattern1, expand=False)
df.loc[df['address_wordlist'].str.contains(pattern1), 'address_wordlist'
      ] = df.loc[df['address_wordlist'].str.contains(pattern1), 'address_wordlist'].str.replace(pattern1, '', regex=True)

<ipython-input-132-161d4ab271fe>:9: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  ] = df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'].str.extract(pattern, expand=False)
<ipython-input-132-161d4ab271fe>:8: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  df.loc[df['address_wordlist'].str.contains(pattern), 'po_box'
<ipython-input-132-161d4ab271fe>:11: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  ] = df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'].str.replace(pattern, '', regex=True)
<ipython-input-132-161d4ab271fe>:10: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  df.loc[df['address_wordlist'].str.contains(pattern), 'address_wordlist'
<ipython-input-132-161d4ab271fe>:15: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  ] 

In [134]:
df.head()

node_id  \
0  24000001   
1  24000002   
2  24000003   
3  24000004   
4  24000005   

                                                                                              address  \
0                                     ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                    SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                      LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3  P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                           LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   

  name countries country_codes       sourceID  \
0  NaN   Bahamas           BHS  Bahamas Leaks   
1  NaN   Bahamas           BHS  Bahamas Leaks   
2  NaN   Bahamas           BHS  Bahamas Leaks   
3  NaN   Bahamas           BHS  Bahamas Leaks   
4  NaN   Bahamas           BHS  Bahamas Leaks   

                                             valid_until note  \
0  The Bahamas Leaks data is current through early 2016.  NaN   
1  The Bahamas Leaks data is current through early 2016.  NaN   
2  The Bahamas Leaks data is current through early 2016.  NaN   
3  The Bahamas Leaks data is current through early 2016.  NaN   
4  The Bahamas Leaks data is current through early 2016.  NaN   

                                                                                  working_address  \
0                                  annex frederick and shirley street po box n4805 nassau bahamas   
1                                       suite e2 union court building po box n8188 nassau bahamas   
2                                         lyford cay house lyford cay po box n7785 nassau bahamas   
3  po box n3708 bahamas financial centre po box n3708 shirley and charlotte street nassau bahamas   
4                             lyford cay house third floor lyford cay po box n3024 nassau bahamas   

                                          address_wordlist address_country  \
0                      annex frederick and shirley street          bahamas   
1                           suite e2 union court building          bahamas   
2                             lyford cay house lyford cay          bahamas   
3   bahamas financial centre  shirley and charlotte street         bahamas   
4                 lyford cay house third floor lyford cay          bahamas   

  address_city        po_box  
0       nassau  po box n4805  
1       nassau  po box n8188  
2       nassau  po box n7785  
3       nassau  po box n3708  
4       nassau  po box n3024

## Cities and islands

#### Cities

- [x] Nassau
- [x] Freeport
- Marsh Harbour
- Governor's Harbour
- Windermere island
- Harbour Island
- Elbow Cay
- Treasure Cay
- Gregory Town
- Spanish Wells

#### Islands

- New Providence
- Paradise Island
- Grand Bahama
- Abaco
- Eleuthera
- South Andros

In [140]:
df.loc[df['address_wordlist'].str.contains(r'(\bfreeport\b)'), ['address_city', 'address_island']] = ['freeport', 'grand bahama']
df['address_wordlist'] = df['address_wordlist'].str.replace('(freeport)', '', regex=True)

gr_bah_pat = r'\bgr\w*\sbah\w+\b'
df.loc[df['working_address'].str.contains(gr_bah_pat), 'address_island'] = 'grand bahama'
df.loc[df['working_address'].str.contains(gr_bah_pat), 'address_wordlist'
      ] = df.loc[df['working_address'].str.contains(gr_bah_pat), 'address_wordlist'].str.replace(gr_bah_pat, '', regex=True)
df.loc[df['working_address'].str.contains(gr_bah_pat), 'address_wordlist'
      ] = df.loc[df['working_address'].str.contains(gr_bah_pat), 'address_wordlist'].str.replace('grand', '', regex=True)

<ipython-input-140-79008a407ff3>:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  df.loc[df['address_wordlist'].str.contains(r'(\bfreeport\b)'), ['address_city', 'address_island']] = ['freeport', 'grand bahama']


In [164]:
df[df['working_address'].str.contains(r'\bnp\b', regex=True)]

node_id  \
163    24000164   
517    24000518   
1595   81031897   
1612   81036376   
1623   81039098   
1733   81080265   
1739   81080985   
1770   81004778   
1923   33000119   
2171  240070266   
2187  240004791   
2220  240452852   

                                                                                              address  \
163                  LYFORD MANOR, WEST BUILDING, WEST BAY ST, P.O. BOX CB-13007, NASSAU, NP, BAHAMAS   
517   3RD FL. BRITISH COLONIAL CENTRE OF COMM, SUITE 304, 1 BAY STREET, SP 63776, NASSAU, NP, BAHAMAS   
1595                                  Sassoon House; Shirley St and Victoria Ave; Nassau; NP; Bahamas   
1612                                      Stanley Eastern Road; P.O. Box 55-5539; Nassau; NP; Bahamas   
1623                             Villa Capulet; Montague Foreshore; PO Box N-8893; Nassau NP; Bahamas   
1733                                  Apartment 2; Coral Harbor; New Providence; NP CB115-21; Bahamas   
1739                            Old Fort Bay; P.O. Box N865; NP N 865 Nassau; New Providence; Bahamas   
1770        PTC Management Limited; 2nd Floor ; Charlotte House; Charlotte Street; Nassau NP; Bahamas   
1923                    LYFORD MANOR, WEST BUILDING, WEST BAY ST PO BOX CB-13007, NASSAU, NP, BAHAMAS   
2171                                   OLD FORT BAY, P.O. BOX N7776, NEW PROVIDENCE NP NASSAU BAHAMAS   
2187                    WEST BAY OCEAN PLACE, NEW PROVIDENCE ISLAND NP N 1836 NASSAU CB133-18 BAHAMAS   
2220                 P O BOX CR56766 STE 875 — NASSAU, BAHAMAS — NEW PROVIDENCE NP CR567-66 - BAHAMAS   

                        name countries country_codes  \
163                      NaN   Bahamas           BHS   
517                      NaN   Bahamas           BHS   
1595           Sassoon House   Bahamas           BHS   
1612    Stanley Eastern Road   Bahamas           BHS   
1623           Villa Capulet   Bahamas           BHS   
1733             Apartment 2   Bahamas           BHS   
1739            Old Fort Bay   Bahamas           BHS   
1770  PTC Management Limited   Bahamas           BHS   
1923                     NaN   Bahamas           BHS   
2171                     NaN   Bahamas           BHS   
2187                     NaN   Bahamas           BHS   
2220                     NaN   Bahamas           BHS   

                                                       sourceID  \
163                                               Bahamas Leaks   
517                                               Bahamas Leaks   
1595                                  Paradise Papers - Appleby   
1612                                  Paradise Papers - Appleby   
1623                                  Paradise Papers - Appleby   
1733                                  Paradise Papers - Appleby   
1739                                  Paradise Papers - Appleby   
1770                                  Paradise Papers - Appleby   
1923               Paradise Papers - Bahamas corporate registry   
2171  Pandora Papers - Alemán, Cordero, Galindo & Lee (Alcogal)   
2187  Pandora Papers - Alemán, Cordero, Galindo & Lee (Alcogal)   
2220                    Pandora Papers - SFM Corporate Services   

                                                  valid_until note  \
163     The Bahamas Leaks data is current through early 2016.  NaN   
517     The Bahamas Leaks data is current through early 2016.  NaN   
1595                     Appleby data is current through 2014  NaN   
1612                     Appleby data is current through 2014  NaN   
1623                     Appleby data is current through 2014  NaN   
1733                     Appleby data is current through 2014  NaN   
1739                     Appleby data is current through 2014  NaN   
1770                     Appleby data is current through 2014  NaN   
1923  Bahamas corporate registry data is current through 2016  NaN   
2171                    Provider data is current through 2017  NaN   
2187                   

In [163]:
df[df['working_address'].str.contains(r'\bnew\s?pr\w+\b', regex=True)]

node_id  \
183    24000184   
235    24000236   
457    24000458   
474    24000475   
530    14000073   
...         ...   
2206  240112487   
2218  240490183   
2220  240452852   
2231  240492015   
2244  240492203   

                                                                                                                                                     address  \
183                                                                             SUITE 1000 NEW PROVIDENCE FINANCIAL CTR., P.O. BOX CR-56766, NASSAU, BAHAMAS   
235                                                                           #4 PINEAPPLE GROVE,OLD FORT BAY, NEW PRODIVENCE, BOX SP-60063, NASSAU, BAHAMAS   
457                                                                                                                 P.O. BOX N-10620, NASSAU, NEW PROVIDENCE   
474                                                                                                                P.O. BOX CB-13265, NASSAU. NEW PROVIDENCE   
530                                                                                   #1 Venetian Villa N492; Old Fort Day; Nassau; New Providence; Bahamas.   
...                                                                                                                                                      ...   
2206                                                                                                              MILLENIUM, NASSAU, NEW PROVIDENCE, BAHAMAS   
2218                                                                                         5 DUNMORE ISLAND, OLD FORT BAY, NASSAU, NEW PROVIDENCE, BAHAMAS   
2220                                                                        P O BOX CR56766 STE 875 — NASSAU, BAHAMAS — NEW PROVIDENCE NP CR567-66 - BAHAMAS   
2231    JPMORGAN TRUST COMPANY (BAHAMAS) LIMITED, 2ND FLOOR, BAHAMAS FINANCIAL CENTRE, SHIRLEY AND CHARLOTTE STREET, NASSAU, N-4899, NEW PROVIDENCE, BAHAMAS   
2244  J.P. MORGAN TRUST COMPANY (BAHAMAS) LIMITED, 2ND FLOOR BAHAMAS FINANCIAL CENTRE, SHIRLEY AND CHARLOTTE STREET, NASSAU, N-4899, NEW PROVIDENCE, BAHAMAS   

     name countries country_codes                                 sourceID  \
183   NaN   Bahamas           BHS                            Bahamas Leaks   
235   NaN   Bahamas           BHS                            Bahamas Leaks   
457   NaN   Bahamas           BHS                            Bahamas Leaks   
474   NaN   Bahamas           BHS                            Bahamas Leaks   
530   NaN   Bahamas           BHS                            Panama Papers   
...   ...       ...           ...                                      ...   
2206  NaN   Bahamas           BHS        Pandora Papers - Alpha Consulting   
2218  NaN   Bahamas           BHS           Pandora Papers - Trident Trust   
2220  NaN   Bahamas           BHS  Pandora Papers - SFM Corporate Services   
2231  NaN   Bahamas           BHS           Pandora Papers - Trident Trust   
2244  NaN   Bahamas           BHS           Pandora Papers - Trident Trust   

                                                valid_until note  \
183   The Bahamas Leaks data is current through early 2016.  NaN   
235   The Bahamas Leaks data is current through early 2016.  NaN   
457   The Bahamas Leaks data is current through early 2016.  NaN   
474   The Bahamas Leaks data is current through early 2016.  NaN   
530         The Panama Papers  data is current through 2015  NaN   
...                                                     ...  ...   
2206                  Provider data is current through 2018  NaN   
2218                  Provider data is current through 2017  NaN   
2220                  Provider data is current through 2012  NaN   
2231                  Provider data is current through 2016  NaN   
2244                  Provider data is current through 2017  NaN   

                                                                                                        

In [162]:
df[df['address'].str.lower().str.contains('grand island', regex=True)]

Empty DataFrame
Columns: [node_id, address, name, countries, country_codes, sourceID, valid_until, note, working_address, address_wordlist, address_country, address_city, po_box, address_island]
Index: []

In [156]:
df

node_id  \
0      24000001   
1      24000002   
2      24000003   
3      24000004   
4      24000005   
...         ...   
2253  240492525   
2254  240492536   
2255  240491733   
2256  240491778   
2257  240491892   

                                                                                                                                          address  \
0                                                                                 ANNEX FREDERICK & SHIRLEY STS, P.O. BOX N-4805, NASSAU, BAHAMAS   
1                                                                                SUITE E-2,UNION COURT BUILDING, P.O. BOX N-8188, NASSAU, BAHAMAS   
2                                                                                  LYFORD CAY HOUSE, LYFORD CAY, P.O. BOX N-7785, NASSAU, BAHAMAS   
3                                              P.O. BOX N-3708 BAHAMAS FINANCIAL CENTRE, P.O. BOX N-3708 SHIRLEY & CHARLOTTE STS, NASSAU, BAHAMAS   
4                                                                       LYFORD CAY HOUSE, 3RD FLOOR, LYFORD CAY, P.O. BOX N-3024, NASSAU, BAHAMAS   
...                                                                                                                                           ...   
2253                                                                      J.P.MORGAN TRUST COMPANY (BAHAMAS) LIMITED, NASSAU, N-4899, ZH, BAHAMAS   
2254                                                                  MONTAGNE STERLINE CENTRE. EAST BAV STREET, NASSAU, COUNTRY BAHAMAS, BAHAMAS   
2255                                                                                     DELTEC HOUSE, LYFORD CAY, PO BOX N-3229, NASSAU, BAHAMAS   
2256                                                  PROVIDENCE HOUSE, HAST WING, EAST HILL STREET, P.O. BOX CB-12399, NASSAU, CB-12399, BAHAMAS   
2257  C/OJ.P. MORGAN TRUST COMPANY (BAHAMAS) LIMITED, BAHAMAS FINANCIAL CENTER 2NTL FLOOR, SHIRLEY AND CHARLOTTE STREETS, NASSAU, N-4899, BAHAMAS   

     name countries country_codes                        sourceID  \
0     NaN   Bahamas           BHS                   Bahamas Leaks   
1     NaN   Bahamas           BHS                   Bahamas Leaks   
2     NaN   Bahamas           BHS                   Bahamas Leaks   
3     NaN   Bahamas           BHS                   Bahamas Leaks   
4     NaN   Bahamas           BHS                   Bahamas Leaks   
...   ...       ...           ...                             ...   
2253  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2254  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2255  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2256  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   
2257  NaN   Bahamas           BHS  Pandora Papers - Trident Trust   

                                                valid_until note  \
0     The Bahamas Leaks data is current through early 2016.  NaN   
1     The Bahamas Leaks data is current through early 2016.  NaN   
2     The Bahamas Leaks data is current through early 2016.  NaN   
3     The Bahamas Leaks data is current through early 2016.  NaN   
4     The Bahamas Leaks data is current through early 2016.  NaN   
...                                                     ...  ...   
2253                  Provider data is current through 2016  NaN   
2254                  Provider data is current through 2016  NaN   
2255                  Provider data is current through 2016  NaN   
2256                  Provider data is current through 2016  NaN   
2257                  Provider data is current through 2017  NaN   

                                                                                                                        working_address  \
0                                                                        annex frederick and shirley street po box n4805 nassau bahamas   
1                                                                 

In [47]:
df['address_wordlist'] = df['address_wordlist'].apply(lambda x: x if not x else x[:-1] if x[-1] in ['the'] else x)